pkg/controllers/workapplier/utils.go

```diff
@@ -0,0 +1,854 @@
+		delete(annotations, corev1.LastAppliedConfigAnnotation)
+		if len(annotations) == 0 {
+			manifestObjCopy.SetAnnotations(nil)
+		} else {
+			manifestObjCopy.SetAnnotations(annotations)
+		}
+	}
+
+	// Remove certain system-managed fields.
+	manifestObjCopy.SetOwnerReferences(nil)
+	manifestObjCopy.SetManagedFields(nil)
+
+	// Remove the read-only fields.
+	manifestObjCopy.SetCreationTimestamp(metav1.Time{})
+	manifestObjCopy.SetDeletionTimestamp(nil)
+	manifestObjCopy.SetDeletionGracePeriodSeconds(nil)
+	manifestObjCopy.SetGeneration(0)
+	manifestObjCopy.SetResourceVersion("")
+	manifestObjCopy.SetSelfLink("")
+	manifestObjCopy.SetUID("")
+
+	// Remove the status field.
+	unstructured.RemoveNestedField(manifestObjCopy.Object, "status")
+
+	// Note: in the Fleet hub agent logic, the system also handles the Service and Job objects
+	// in a special way, so as to remove certain fields that are set by the hub cluster API
+	// server automatically; for the Fleet member agent logic here, however, Fleet assumes
+	// that if these fields are set, users must have set them on purpose, and they should not
+	// be removed. The difference comes to the fact that the Fleet member agent sanitization
+	// logic concerns only the enveloped objects, which are free from any hub cluster API
+	// server manipulation anyway.
+
+	return manifestObjCopy
+}
+
+// shouldEnableOptimisticLock checks if optimistic lock should be enabled given an apply strategy.
+func shouldEnableOptimisticLock(applyStrategy *fleetv1beta1.ApplyStrategy) bool {
+	// Optimistic lock is enabled if the apply strategy is set to IfNotDrifted.
+	return applyStrategy.WhenToApply == fleetv1beta1.WhenToApplyTypeIfNotDrifted
+}
+
+// shouldPerformPreApplyDriftDetection checks if pre-apply drift detection should be performed.
+func shouldPerformPreApplyDriftDetection(manifestObj, inMemberClusterObj *unstructured.Unstructured, applyStrategy *fleetv1beta1.ApplyStrategy) (bool, error) {
+	// Drift detection is performed before the apply op if (and only if):
+	// * Fleet reports that the manifest has been applied before (i.e., inMemberClusterObj exists); and
+	// * The apply strategy dictates that an apply op should only run if there is no
+	//   detected drift; and
+	// * The hash of the manifest object is consistently with that bookkept in the live object
+	//   annotations (i.e., the manifest object has been applied before).
+	if applyStrategy.WhenToApply != fleetv1beta1.WhenToApplyTypeIfNotDrifted || inMemberClusterObj == nil {
+		// A shortcut to save some overhead.
+		return false, nil
+	}
+
+	cleanedManifestObj := discardFieldsIrrelevantInComparisonFrom(manifestObj)
+	manifestObjHash, err := resource.HashOf(cleanedManifestObj.Object)
+	if err != nil {
+		return false, err
+	}
+
+	inMemberClusterObjLastAppliedManifestObjHash := inMemberClusterObj.GetAnnotations()[fleetv1beta1.ManifestHashAnnotation]
+	return manifestObjHash == inMemberClusterObjLastAppliedManifestObjHash, nil
```

Comment:
why check if the manifest matches?

Reply:
Hi Ryan! If the manifest itself has been updated (i.e., a new version of the manifest has become available), there is no need to do drift detection anymore; we will simply apply the new version.